front/src/applications/stdcm/utils/index.ts

```diff
@@ -43,3 +44,50 @@ export const extractMarkersInfo = (pathSteps: StdcmPathStep[]): MarkerInformatio
 
     return acc;
   }, []);
+
+/**
+ * Merges segments from multiple API calls, prioritizing earlier calls (more precise).
+ * Then compresses successive segments with the same train name.
+ */
+export const mergeSimilarTrainSegments = (
+  segmentGroups: PostSimilarTrainsApiResponse['similar_trains'][]
+): PostSimilarTrainsApiResponse['similar_trains'] => {
+  const segmentMap = new Map<string, PostSimilarTrainsApiResponse['similar_trains'][number]>();
+  const referenceOrder: string[] = [];
+
+  // Use the first group's order as the reference
+  // starting with the first group (to keep its relative order)
+  for (const segments of segmentGroups) {
+    for (const segment of segments) {
+      const key = `${segment.begin}-${segment.end}`;
+
+      if (!segmentMap.has(key)) {
+        referenceOrder.push(key);
+      }
+
+      const existing = segmentMap.get(key);
+
+      // Add segment if it doesn't exist, or if existing is null but this one isn't
+      if (!existing || (existing.train === null && segment.train !== null)) {
+        segmentMap.set(key, segment);
+      }
+    }
+  }
+
+  // Convert to array in order, merging adjacent segments with same train
+  return referenceOrder.reduce<PostSimilarTrainsApiResponse['similar_trains']>((acc, key) => {
+    const segment = segmentMap.get(key)!;
+    const lastSegment = acc[acc.length - 1];
+    const lastTrain = lastSegment?.train?.train_name ?? null;
+    const currentTrain = segment.train?.train_name ?? null;
+
+    // Merge if adjacent and same train
+    if (lastSegment?.end === segment.begin && lastTrain === currentTrain) {
+      lastSegment.end = segment.end;
+    } else {
+      acc.push({ ...segment });
```

Comment:
You don't need to recreate a new object here :)
```suggestion
      acc.push(segment);
```

Reply:
Done